helix-core/src/flock.rs

```diff
@@ -0,0 +1,123 @@
+use std::fs::{File, OpenOptions};
+use std::io::Result;
+use std::path::Path;
+
+pub struct FileLock {
+    file: File,
+    shared: bool,
+}
+
+impl FileLock {
+    pub fn exclusive<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file = Self::open_lock(path)?;
+        Ok(Self {
+            file,
+            shared: false,
+        })
+    }
+
+    pub fn shared<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file = Self::open_lock(path)?;
+        Ok(Self { file, shared: true })
+    }
+
+    pub fn get(&self) -> Result<&File> {
+        self.lock()?;
+        Ok(&self.file)
+    }
+
+    pub fn get_mut(&mut self) -> Result<&mut File> {
+        self.lock()?;
+        Ok(&mut self.file)
+    }
+
+    fn lock(&self) -> Result<()> {
```

Comment:
Does it make sense to make this async, or will this unnecessarily complicate things?

Reply:
I don't actually think we need a file lock at all. 

@kirawi since #5805 has landed helix already checks that there is no external contention so if we are saving the file it is also save to write the undofile. I don't think we need a lock anymore. I think the main reason you had a file lock to begin with was to protect the shared index file if  am not mistaken but since that has been removed, I think the flock can be removed too (which would make this PR smaller)

Reply:
I was also thinking about in case multiple Helix instances would write to the same undo file at the same time. Theoretically, isn't it possible for multiple instances to pass the mtime check? Although, I can't imagine anyone writing to the same file from multiple instances in the first place.